In [21]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder

In [22]:
train_features = pd.read_csv("/content/features.csv")
train_labels = pd.read_csv("/content/labels.csv")

In [23]:
print(train_features)
# trains_features

      Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  95

In [24]:
print(train_labels)
# trains_labels


        Target
0     Graduate
1     Graduate
2     Graduate
3     Graduate
4      Dropout
...        ...
3091  Graduate
3092  Graduate
3093  Graduate
3094   Dropout
3095  Graduate

[3096 rows x 1 columns]




Berikut adalah penjelasan untuk setiap kolom pada dataset:

    Unnamed: 0: Kolom indeks yang dihasilkan saat menyimpan CSV.
    Student Id: ID unik untuk setiap siswa.
    Marital status: Status perkawinan siswa.
    Application mode: Mode aplikasi yang digunakan siswa.
    Application order: Urutan aplikasi siswa.
    Course: Kode kursus yang diambil siswa.
    Daytime/evening attendance: Kehadiran siswa (siang/malam).
    Previous qualification: Kualifikasi sebelumnya dari siswa.
    Previous qualification (grade): Nilai kualifikasi sebelumnya.
    Nacionality: Kebangsaan siswa.
    Mother's qualification: Kualifikasi ibu siswa.
    Father's qualification: Kualifikasi ayah siswa.
    Mother's occupation: Pekerjaan ibu siswa.
    Father's occupation: Pekerjaan ayah siswa.
    Admission grade: Nilai penerimaan siswa.
    Displaced: Apakah siswa mengungsi.
    Educational special needs: Apakah siswa memiliki kebutuhan pendidikan khusus.
    Debtor: Apakah siswa memiliki tunggakan.
    Tuition fees up to date: Apakah biaya kuliah siswa sudah lunas.
    Gender: Jenis kelamin siswa.
    Scholarship holder: Apakah siswa penerima beasiswa.
    Age at enrollment: Usia siswa saat mendaftar.
    International: Apakah siswa internasional.
    Curricular units 1st sem (credited): Unit kurikuler semester 1 yang dikreditkan.
    Curricular units 1st sem (enrolled): Unit kurikuler semester 1 yang terdaftar.
    Curricular units 1st sem (evaluations): Unit kurikuler semester 1 yang dievaluasi.
    Curricular units 1st sem (approved): Unit kurikuler semester 1 yang disetujui.
    Curricular units 1st sem (grade): Nilai unit kurikuler semester 1.
    Curricular units 1st sem (without evaluations): Unit kurikuler semester 1 tanpa evaluasi.
    Curricular units 2nd sem (credited): Unit kurikuler semester 2 yang dikreditkan.
    Curricular units 2nd sem (enrolled): Unit kurikuler semester 2 yang terdaftar.
    Curricular units 2nd sem (evaluations): Unit kurikuler semester 2 yang dievaluasi.
    Curricular units 2nd sem (approved): Unit kurikuler semester 2 yang disetujui.
    Curricular units 2nd sem (grade): Nilai unit kurikuler semester 2.
    Curricular units 2nd sem (without evaluations): Unit kurikuler semester 2 tanpa evaluasi.
    Unemployment rate: Tingkat pengangguran.
    Inflation rate: Tingkat inflasi.
    GDP: Produk Domestik Bruto.
    Target: Status siswa (Lulus, Putus Sekolah, Terdaftar).

In [25]:
df = pd.concat([train_features, train_labels], axis=1)
df

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   
...                                   ...   
3091                                  6.0   
3092                                  6.0   
3093                                  8.0   
3094                                  8.0   
3095                                  6.0   

      Curricular units 2nd sem (evaluations)  \
0                                        6.0   
1                                        0.0   
2                                        8.0   
3                                        7.0   
4                                        NaN   
...                                      ...   
3091                                     9.0   
3092             

3. Menampilkan tipe data dari setiap kolom yang terdapat dalam dataset

In [26]:
df.dtypes

Unnamed: 0                                          int64
Student Id                                          int64
Marital status                                     object
Application mode                                  float64
Application order                                 float64
Course                                            float64
Daytime/evening attendance\t                       object
Previous qualification                            float64
Previous qualification (grade)                    float64
Nacionality                                        object
Mother's qualification                            float64
Father's qualification                            float64
Mother's occupation                               float64
Father's occupation                               float64
Admission grade                                   float64
Displaced                                         float64
Educational special needs                         float64
Debtor                                            float64
Tuition fees up to date                           float64
Gender                                            float64
Scholarship holder                                float64
Age at enrollment                                 float64
International                                     float64
Curricular units 1st sem (credited)               float64
Curricular units 1st sem (enrolled)               float64
Curricular units 1st sem (evaluations)            float64
Curricular units 1st sem (approved)               float64
Curricular units 1st sem (grade)                  float64
Curricular units 1st sem (without evaluations)    float64
Curricular units 2nd sem (credited)               float64
Curricular units 2nd sem (enrolled)               float64
Curricular units 2nd sem (evaluations)            float64
Curricular units 2nd sem (approved)               float64
Curricular units 2nd sem (grade)                  float64
Curricular units 2nd sem (without evaluations)    float64
Unemployment rate                                 float64
Inflation rate                                    float64
GDP                                               float64
Target                                             object
dtype: object

4. menampilkan statistik deskriptif dari dataset dan berikan analisis singkat mengenai hasilnya

In [27]:
df.describe()

Unnamed: 0   Student Id  Application mode  Application order  \
count  3096.000000  3096.000000       2924.000000        2988.000000   
mean   1547.500000  5470.623062         18.785568           1.734940   
std     893.882543  2582.483019         17.532674           1.327527   
min       0.000000  1000.000000          1.000000           1.000000   
25%     773.750000  3288.500000          1.000000           1.000000   
50%    1547.500000  5441.500000         17.000000           1.000000   
75%    2321.250000  7712.250000         39.000000           2.000000   
max    3095.000000  9997.000000         53.000000           9.000000   

            Course  Previous qualification  Previous qualification (grade)  \
count  2944.000000             2973.000000                     3011.000000   
mean   8817.929348                4.685503                      132.647293   
std    2148.407432               10.439709                       13.322662   
min      33.000000                1.000000                       95.000000   
25%    9085.000000                1.000000                      124.000000   
50%    9238.000000                1.000000                      133.100000   
75%    9556.000000                1.000000                      140.000000   
max    9991.000000               43.000000                      190.000000   

       Mother's qualification  Father's qualification  Mother's occupation  \
count             2951.000000             3000.000000          3034.000000   
mean                19.470010               22.408667            11.495715   
std                 15.636952               15.291008            28.173386   
min                  1.000000                1.000000             0.000000   
25%                  2.000000                3.000000             4.000000   
50%                 19.000000               19.000000             5.000000   
75%                 37.000000               37.000000             9.000000   
max                 44.000000               44.000000           194.000000   

       ...  Curricular units 1st sem (without evaluations)  \
count  ...                                     3023.000000   
mean   ...                                        0.133311   
std    ...                                        0.666707   
min    ...                                        0.000000   
25%    ...                                        0.000000   
50%    ...                                        0.000000   
75%    ...                                        0.000000   
max    ...                                       10.000000   

       Curricular units 2nd sem (credited)  \
count                          3035.000000   
mean                              0.516639   
std                               1.840537   
min                               0.000000   
25%                               0.000000   
50%                               0.000000   
75%                               0.000000   
max                              19.000000   

       Curricular units 2nd sem (enrolled)  \
count                          2997.000000   
mean                              6.208876   
std                               2.185127   
min                               0.000000   
25%                               5.000000   
50%                               6.000000   
75%                               7.000000   
max                              23.000000   

       Curricular units 2nd sem (evaluations)  \
count                             2889.000000   
mean                                 8.061613   
std                                  3.942850   
min                                  0.000000   
25%                                  6.000000   
50%                                  8.000000   
75%                                 10.000000   
max                                 27.000000   

       Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
count                          2945.000000

Analisis deskriptif dari dataset:

Output dari `df.describe()` memberikan ringkasan statistik untuk kolom-kolom numerik dalam dataset. Beberapa poin penting yang dapat diamati adalah:

*   **Count:** Menunjukkan jumlah nilai non-null untuk setiap kolom. Terlihat bahwa ada beberapa kolom dengan jumlah nilai yang kurang dari total baris (3096), menandakan adanya nilai yang hilang (NaN).
*   **Mean:** Rata-rata nilai untuk setiap kolom. Ini memberikan gambaran tentang nilai pusat dari distribusi data.
*   **Std:** Standar deviasi, mengukur sebaran data dari rata-rata. Nilai standar deviasi yang tinggi menunjukkan sebaran data yang lebih luas.
*   **Min dan Max:** Menunjukkan nilai minimum dan maksimum dalam setiap kolom. Ini membantu mengidentifikasi rentang nilai data.
*   **25%, 50%, 75% (Quartiles):** Menunjukkan kuartil pertama, median (kuartil kedua), dan kuartil ketiga. Ini memberikan wawasan tentang distribusi data dan keberadaan outlier.

Dari analisis ini, kita dapat melihat adanya variasi yang signifikan dalam beberapa kolom, serta keberadaan nilai yang hilang yang perlu ditangani lebih lanjut.

5. slicing data


In [28]:
# menampilkan 20 baris awal dari dataset
df.head(20)

Unnamed: 0  Student Id Marital status  Application mode  \
0            0        6634         single               1.0   
1            1        6459         single               1.0   
2            2        2238         single              42.0   
3            3        2479         single               1.0   
4            4        4260         single              44.0   
5            5        3544        married              39.0   
6            6        4956         single              39.0   
7            7        6758         single               1.0   
8            8        2484         single               1.0   
9            9        9845         single              39.0   
10          10        3690         single               1.0   
11          11        2585         single               1.0   
12          12        5241         single               1.0   
13          13        2972         single               1.0   
14          14        2003        married              39.0   
15          15        3007         single               7.0   
16          16        9857         single              17.0   
17          17        3327         single               NaN   
18          18        5235         single              17.0   
19          19        3969         single               1.0   

    Application order  Course Daytime/evening attendance\t  \
0                 3.0  9773.0                          day   
1                 1.0     NaN                          day   
2                 1.0  9500.0                          day   
3                 1.0     NaN                          day   
4                 1.0     NaN                          day   
5                 1.0  9853.0                          day   
6                 1.0  9085.0                          day   
7                 4.0  9670.0                          day   
8                 1.0  9773.0                          day   
9                 1.0  9670.0                          day   
10                1.0  9670.0                          day   
11                1.0  9070.0                          day   
12                1.0   171.0                          day   
13                3.0  9500.0                          day   
14                1.0  8014.0                          NaN   
15                1.0  9119.0                          day   
16                1.0  9119.0                          day   
17                1.0  9003.0                          day   
18                1.0   171.0                          day   
19                1.0  9085.0                          day   

    Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                      1.0                             NaN  Portuguese  ...   
1                      1.0                           136.0  Portuguese  ...   
2                      1.0                           120.0  Portuguese  ...   
3                      1.0                           141.0  Portuguese  ...   
4                     39.0                           120.0     Spanish  ...   
5                      1.0                             NaN  Portuguese  ...   
6                      1.0                           133.1         NaN  ...   
7                      1.0                           130.0  Portuguese  ...   
8                      1.0                           135.0  Portuguese  ...   
9                      1.0                           120.0  Portuguese  ...   
10                     1.0                           133.1  Portuguese  ...   
11                     1.0                           117.0  Portuguese  ...   
12                     1.0                           162.0  Portuguese  ...   
13                     1.0                           135.0  Portuguese  ...   
14                     NaN                           120.0  Portuguese  ...   
15                     1.0                           140.8  Portuguese  ...   
16                     1.0        

In [29]:
# menampilkan 20 baris terakhir
df.tail(20)

Unnamed: 0  Student Id Marital status  Application mode  \
3076        3076        6757         single              17.0   
3077        3077        5750         single               1.0   
3078        3078        2599         single               1.0   
3079        3079        3249         single              17.0   
3080        3080        3366       divorced              39.0   
3081        3081        3462         single              39.0   
3082        3082        9194         single              17.0   
3083        3083        5574         single               1.0   
3084        3084        8311         single               1.0   
3085        3085        2304         single              17.0   
3086        3086        6615         single              43.0   
3087        3087        3759         single               1.0   
3088        3088        8085         single              51.0   
3089        3089        1573         single              17.0   
3090        3090        9737         single              51.0   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
3076                NaN  9254.0                          day   
3077                2.0  9500.0                          day   
3078                1.0  9085.0                          day   
3079                1.0  9238.0                          day   
3080                1.0  9003.0                          day   
3081                1.0   171.0                          day   
3082                5.0  9147.0                          day   
3083                1.0  9500.0                          day   
3084                2.0  9773.0                          day   
3085                1.0  9238.0                          day   
3086                1.0  9556.0                          day   
3087                1.0   171.0                          day   
3088                1.0  9670.0                          day   
3089                2.0   171.0                          day   
3090                1.0  9070.0                          day   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
3076                     1.0                           122.0  Portuguese  ...   
3077                     1.0                           140.0  Portuguese  ...   
3078                     1.0                           131.0  Portuguese  ...   
3079                     1.0                           120.0  Portuguese  ...   
3080                    12.0                           133.1  Portuguese  ...   
3081                     1.0                           147.0  Portuguese  ...   
3082                     1.0                           131.0  Portuguese  ...   
3083                     1.0                           140.0  Portuguese  ...   
3084                     1.0                           130.0  Portuguese  ...   
3085                     NaN                           139.0  Portuguese  ...   
3086                     1.0                           149.0  Portuguese  ...   
3087                     3.0                           116.0  Portuguese  ...   
3088                     1.0                           133.1  Portuguese  ...   
3089                     1.0                           142.0  Portuguese  ...   
3090                     1.0                           143.0  Portuguese  ...   

In [30]:
# menampilkan baris dari index 80-100
display(df.iloc[80:101])

Unnamed: 0  Student Id Marital status  Application mode  \
80           80        1238         single              17.0   
81           81        4121         single              39.0   
82           82        2890         single               1.0   
83           83        4358         single               1.0   
84           84        4896         single               NaN   
85           85        4735         single              18.0   
86           86        1047         single               1.0   
87           87        7430         single              39.0   
88           88        8527         single               1.0   
89           89        5420         single              43.0   
90           90        5625         single              44.0   
91           91        8990            NaN               1.0   
92           92        9143        married              44.0   
93           93        1944        married               7.0   
94           94        3666         single               1.0   
95           95        6960         single              17.0   
96           96        7620         single               1.0   
97           97        6083         single              42.0   
98           98        5525         single              17.0   
99           99        6029         single              17.0   
100         100        9749         single              39.0   

     Application order  Course Daytime/evening attendance\t  \
80                 1.0  9670.0                          day   
81                 1.0  9003.0                          day   
82                 1.0   171.0                          day   
83                 1.0  9147.0                          day   
84                 1.0  9991.0                      evening   
85                 1.0  9500.0                          day   
86                 1.0  9773.0                          day   
87                 2.0  9556.0                          day   
88                 2.0  9500.0                          day   
89                 1.0     NaN                          day   
90                 1.0  9003.0                          day   
91                 1.0  9500.0                          day   
92                 1.0  9085.0                          day   
93                 1.0  9070.0                          day   
94                 3.0  9070.0                          day   
95                 3.0  9119.0                          day   
96                 1.0  9500.0                          day   
97                 1.0  9500.0                          day   
98                 5.0  9238.0                          day   
99                 1.0  9773.0                          day   
100                1.0  9085.0                          day   

     Previous qualification  Previous qualification (grade) Nacionality  ...  \
80                      1.0                           155.0  Portuguese  ...   
81                      1.0                             NaN  Portuguese  ...   
82                      1.0                           159.0         NaN  ...   
83                      1.0                           150.0   Ukrainian  ...   
84                      1.0                           140.0  Portuguese  ...   
85                      1.0                           144.0  Portuguese  ...   
86                      1.0                           117.0  Portuguese  ...   
87                     19.0                           133.1  Portuguese  ...   
88                      1.0                           145.0  Portuguese  ...   
89                      1.0                           121.0  Portuguese  ...   
90                     39.0                           140.0  Portuguese  ...   
91                      1.0                           141.0  Portuguese  ...   
92                     39.0                           160.0  Portuguese  ...   
93                     43.0                           140.0  Portuguese  ...   
94       

6. tangani nilai yang hilang (NaN) menggunakan salah satu metode berikut:

In [31]:
#menampilkan data sebelum di imputasi
print(df.isnull().sum())
print("\ntotal nilai hilang sebelum imputasi: ", df.isnull().sum().sum())


Unnamed: 0                                          0
Student Id                                          0
Marital status                                     84
Application mode                                  172
Application order                                 108
Course                                            152
Daytime/evening attendance\t                       80
Previous qualification                            123
Previous qualification (grade)                     85
Nacionality                                       185
Mother's qualification                            145
Father's qualification                             96
Mother's occupation                                62
Father's occupation                               144
Admission grade                                   141
Displaced                                         122
Educational special needs                         119
Debtor                                            116
Tuition fees up to date     

In [32]:
#ubah kategori menjadi angka (karna imputasi hanya bisa menggunakan angka)
df_encoded = df.copy()
for col in df_encoded.select_dtypes(include=['object']):
    df_encoded[col] = df_encoded[col].astype('category').cat.codes

In [33]:
#Buat dan jalankan SimpleImputer ---
# Pilihan strategi:
# ganti nan dengan rata - rata
imputer = SimpleImputer(strategy='mean')

In [34]:
# Lakukan imputasi
df_imputed = pd.DataFrame(imputer.fit_transform(df_encoded),
                          columns=df_encoded.columns)

In [35]:
#menampilkan jumlah data hilang sesudah imputasi
print(df_imputed.isnull().sum())
print("\ntotal nilai hilang sesudah imputasi: ",df_imputed.isnull().sum().sum())


Unnamed: 0                                        0
Student Id                                        0
Marital status                                    0
Application mode                                  0
Application order                                 0
Course                                            0
Daytime/evening attendance\t                      0
Previous qualification                            0
Previous qualification (grade)                    0
Nacionality                                       0
Mother's qualification                            0
Father's qualification                            0
Mother's occupation                               0
Father's occupation                               0
Admission grade                                   0
Displaced                                         0
Educational special needs                         0
Debtor                                            0
Tuition fees up to date                           0
Gender      


8. tampilkan kembali dataset yang telah melalui seluruh proses preprocessing

In [36]:
#tampilkan baris pertama sebelum imputasi
print(df.head)

<bound method NDFrame.head of       Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day 

In [37]:
#menampilkan beberapa baris sesudah imputasi
print(df_imputed.head())

   Unnamed: 0  Student Id  Marital status  Application mode  \
0         0.0      6634.0             4.0               1.0   
1         1.0      6459.0             4.0               1.0   
2         2.0      2238.0             4.0              42.0   
3         3.0      2479.0             4.0               1.0   
4         4.0      4260.0             4.0              44.0   

   Application order       Course  Daytime/evening attendance\t  \
0                3.0  9773.000000                           0.0   
1                1.0  8817.929348                           0.0   
2                1.0  9500.000000                           0.0   
3                1.0  8817.929348                           0.0   
4                1.0  8817.929348                           0.0   

   Previous qualification  Previous qualification (grade)  Nacionality  ...  \
0                     1.0                      132.647293         13.0  ...   
1                     1.0                      136.000000   